# Decombinator-Tools

This repository contains scripts that may be helpful when working with the [Decombinator](https://github.com/innate2adaptive/Decombinator) software.

<h3 id="top">Navigation:</h3>

* [Test Data Generator](#test-data-generator)
* [Log Summary](#log-summary)
* [DCR to Gene Name](#dcr-to-gene)
* [Recipes](#recipes)

---
<h1 id="test-data-generator">Test Data Generator</h1>

The automatic test data generator can be used to produce customised FASTQ files to use as test input data for use with the [Decombinator Script](https://github.com/innate2adaptive/Decombinator#decombinator). This script should be run using **Python 2**.

Generic TCR alpha and beta chain FASTQ files for Decombinator can be generated by simply running:
```
python test-data-generator.py
```
Separate FASTQ files containing alpha and beta TCR sequences will be saved to a newly created `TestData` directory. Each file will contain 100 FASTQ reads each containing a V and J tag. The direction of these reads by default is in the reverse direction (which is also the default for Decombinator). By default, these sequences will include dummy barcodes (with a default barcode length of 42).

The id of each read in the FASTQ file contains information regarding which chain is present in the read, the direction of the read, whether the read contains one (SingleTag) or two (DualTag) tags, and a read-id.

## Customisability

The test data generator can be provided a number of input arguments to customise the output test data. Information on these arguments can be found be running:
```
python test-data-generator.py -h
```
#### Input Arguments
| Argument | Description | 
|:---:|---|
| `-d`  | change the name of the output directory for the test data |
| `-rl` | change the read length of the output sequences (before barcoding) |
| `-n`  | change the total number of output reads per chain |
| `-v`  | choose whether to produce dual tag or single tag output data |
| `-mc` | produces a single output file of mixed alpha and beta chain TCRs |
| `-nbc`| choose to produce non-barcoded data |
| `-bl` | change the length of the attached barcode |
| `-it` | provide a percentage value for how many of the reads contain TCR sequences |
| `-ie` | provide a percentage value for how many tags contain one "sequencing" error |
| `-mt` | provide a percentage value for how many alpha and beta tags will have an overlapping match in single tag mode |
| `-tf` | direct the script to the local directory where tag files are stored. Useful for offline work. By default the script will first attempt to download the tags from the [Decombinator-Tags-FASTAs](https://github.com/innate2adaptive/Decombinator-Tags-FASTAs) directory. |
| `-or` | change the orientation of the output reads, 'reverse', 'forward', or 'both' (50% forward and 50% reverse) |

#### Examples

* To create a file of 250 non-barcoded, mixed alpha and beta chain reads of 160 bp in length, stored in a directory called `MyData` run:
```
python test-data-generator.py -nbc -mc -n 250 -rl 160 -d MyData
```

* To create alpha and beta files for Single Tag Decombinator, in the forward direction, with 40% of sequences containing a sequencing error (1 bp substitution), run:

```
python test-data-generator.py -v s -or forward -ie 40
```
<h1 id="test-data-generator">Log Summary</h1>

This script will generate a summary csv file of all the data stored in the `Logs` folder within the Decombinator repository. This script is should be run in **Python 3**.

How to run:
```
python log-summary.py /path/to/LogsFolder/ /path/to/outfile.csv
```
The output csv file contains the following fields:

| sample |
| NumberReadsInput |
| NumberReadsDecombined |
| PercentReadsDecombined |
| UniqueDCRsPassingFilters |
| TotalDCRsPassingFilters |
| PercentDCRPassingFilters(withbarcode) |
| UniqueDCRsPostCollapsing |
| TotalDCRsPostCollapsing |
| PercentUniqueDCRsKept |
| PercentTotalDCRsKept |
| AverageInputTCRAbundance |
| AverageOutputTCRAbundance |
| AverageRNAduplication |

<h1 id="dcr-to-gene">DCR to Gene Name</h1>

The first five comma-delimited fields in a Decombinator or Collapsinator output file are referred to as the DCR identifier. The first two of these five fields are integer indices that correspond to V and J genes.

The `DCRtoGeneName` script reads in a file containing a DCR identifier and converts these first two V and J indices to the real gene name. The output (path and) file name is the same as the input name, but has prefixes the file extension with `tr_` e.g. `example.n12.gz` will convert to `example.tr_n12.gz`.

How to run:
```
python DCRtoGeneName.py -in path/to/input_alpha_file.n12.gz 
```

If chain name is not present in the input file name, it should be supplied by the user, e.g. : 
```
python DCRtoGeneName.py -in path/to/input_file.freq -c beta 
```

The default species for this script is assumed to be human, but can be changed with the `-sp` argument, e.g. `-sp mouse`. A full list of additional arguments can be viewed by running:
```
python DCRtoGeneName.py -h
```

<h1 id="recipes">Recipes</h1>

The recipes folder contains a number of bash scripts that can be used or modified as a shorthand to run over multiple input files in a directory.

* [Decombinator.sh](#decombinator.sh)
* [Collapsinator.sh](#collapsinator.sh)

<h3 id="decombinator.sh">Decombinator.sh</h3>

This script will run Decombinator for all `.fq` and `.fq.gz` files in the Decombinator directory. Note that this script assumes the Decombinator and Decombinator-Tools repositories are located in the same parent directory. Otherwise a path to the Decombinator repository should be supplied. Additionally, the `Logs` folder for these files will be located in the working directory from which the script is run - to keep log files in the Decombinator repository, follow step 4 below.

# How to run:
1. from recipes folder:
```
./Decombinator.sh
```
or
```  
source Decombinator.sh
```
2. from Decombinator-Tools folder:
```
./recipes/Decombinator.sh
```
or
```
source recipes/Collaspinsator.sh
```
3. from Decombinator folder:
```
source /path/to/Decombinator-Tools/recipes/Decombinator.sh
```
4. Supplying path to Decombinator repository
```
source Decombinator.sh /path/to/Decombinator
```
<h3 id="collapsinator.sh">Collapsinator.sh</h3>

This script will run Collapsinator for all `.n12.gz` files in the Decombinator directory in parallel. Note that this script assumes the Decombinator and Decombinator-Tools repositories are located in the same parent directory. Otherwise a path to the Decombinator repository should be supplied. Additionally, the `Logs` folder for these files will be located in the working directory from which the script is run - to keep log files in the Decombinator repository, follow step 4 below.

# How to run:
1. from recipes folder:
```
./Collapsinator.sh
```
or
```  
source Collapsinator.sh
```
2. from Decombinator-Tools folder:
```
./recipes/Collapsinator.sh
```
or
```
source recipes/Collaspinsator.sh
```
3. from Decombinator folder:
```
source /path/to/Decombinator-Tools/recipes/Collapsinator.sh
```
4. Supplying path to Decombinator repository
```
source Collapsinator.sh /path/to/Decombinator
```